internal/indexer/symbols.go

```diff
@@ -0,0 +1,225 @@
+	document := &protocol.RangeBasedDocumentSymbol{}
+
+	// Visit each node the AST, emitting `RangeBasedDocumentSymbol` for each node we're
+	// interested in.
+	var firstErr error
+	var visitor ast.Visitor
+	visitor = visitorFunc(func(n ast.Node) (w ast.Visitor) {
+		switch node := n.(type) {
+		//case *ast.TypeSpec:
+		// Struct, interace, and other type definitions.
+		case *ast.GenDecl:
+			// Constants, variables
+			if node.Tok == token.CONST || node.Tok == token.VAR {
+				symbolKind := protocol.Constant
+				if node.Tok == token.VAR {
+					symbolKind = protocol.Variable
+				}
+				for _, name := range specNames(node.Specs) {
+					if name == "_" {
+						// Ignore `const _ T = foo` assignments: they cannot be referred to at all.
+						continue
+					}
+					child, err := s.indexSymbol(p, f, name, "", "", symbolKind, declNamePos(node, name))
+					if err != nil {
+						firstErr = errors.Wrap(err, "symbol "+name)
+						return nil
+					}
+					document.Children = append(document.Children, child)
+				}
+			}
+			//case *ast.FuncDecl:
+			// Functions, methods
+		}
+		return visitor
+	})
+	ast.Walk(visitor, f)
+	if firstErr != nil {
+		return firstErr
+	}
+
+	// Emit the edge for the document itself.
+	document.ID = s.i.emitRangeForSymbol(p.Fset.Position(f.Name.Pos()), len(f.Name.Name), &protocol.RangeTag{
+		Type:   "definition",
+		Text:   p.PkgPath,
+		Detail: "file doc text", // TODO(slimsag): file.Doc.Text(),
+		Kind:   protocol.Package,
+		// TODO(slimsag): should exported status depend on whether or not file contains exported items?
+		Tags: []protocol.SymbolTag{protocol.Exported},
+	})
+	filename := p.Fset.Position(f.Pos()).Filename
+	doc := s.i.documents[filename]
+	if doc == nil {
+		return fmt.Errorf("(unexpected) document not found after indexing: %s", filename)
+	}
+	results := []*protocol.RangeBasedDocumentSymbol{document}
+	resultID := s.i.emitter.EmitDocumentSymbolResult(results)
+	s.i.emitter.EmitDocumentSymbolEdge(resultID, doc.DocumentID)
+	return nil
+}
+
+var n = 0
+
+// indexSymbol emits the `RangeBasedDocumentSymbol` corresponding to the given singular symbol.
+// e.g., a variable, constant, function, method, type, field, etc.
+//
+// * name is the name of the symbol
+// * recv is the name of the Go receiver (e.g. struct), or an empty string.
+// * container is the name of the container (e.g. struct or interface name), if any.
+//
+func (s *symbolIndexer) indexSymbol(
+	p *packages.Package,
+	f *ast.File,
+	name, recv, container string,
+	kind protocol.SymbolKind,
+	pos token.Pos,
+) (*protocol.RangeBasedDocumentSymbol, error) {
+	sym := &protocol.RangeBasedDocumentSymbol{}
+	// If the below is changed to 2, all tests pass. If it's 3, then:
+	//
+	// --- FAIL: TestIndexer (1.21s)
+	//     --- FAIL: TestIndexer/check_wg_references (0.00s)
+	//         indexer_test.go:108: incorrect reference count. want=4 have=1
+	// FAIL
+	// FAIL	github.com/sourcegraph/lsif-go/internal/indexer	18.149s
+	// FAIL
+	//
```

Comment:
Could use help understanding what is going wrong here; not sure why this causes an implicit change in behavior in reference generation

Reply:
This turned out to be due to https://github.com/sourcegraph/lsif-go/pull/142